File: js/cartas.js
// ============================================================
//  CARTAS, MAZO MÚLTIPLE (SHOE) Y CONTEO
//  Aquí vive todo lo relacionado con las cartas: crear el shoe,
//  barajar, repartir, calcular puntos, conteo Hi-Lo y
//  probabilidades. NO toca el DOM (eso es trabajo de ui.js).
// ============================================================

const PALOS = ["♠", "♥", "♦", "♣"];
const VALORES = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"];

let shoe = [];          // El "zapato": varias barajas mezcladas
let totalShoe = 0;      // Cuántas cartas tenía el shoe recién barajado
let conteoCorrido = 0;  // Conteo Hi-Lo "corrido" (running count)

// Crea el shoe con N barajas y lo baraja
function crearShoe() {
  shoe = [];
  for (let b = 0; b < config.numBarajas; b++) {
    for (const palo of PALOS) {
      for (const valor of VALORES) {
        shoe.push({ palo, valor });
      }
    }
  }
  barajar(shoe);
  totalShoe = shoe.length;
  conteoCorrido = 0;
}

// Algoritmo Fisher-Yates (mezcla justa)
function barajar(m) {
  for (let i = m.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [m[i], m[j]] = [m[j], m[i]];
  }
  return m;
}

// ¿Conviene volver a barajar? (penetración del 25%, como en casinos)
function necesitaBarajar() {
  return shoe.length < totalShoe * 0.25;
}

// Saca la carta de arriba del shoe (si se acaba, crea uno nuevo)
function sacarCarta() {
  if (shoe.length === 0) crearShoe();
  return shoe.pop();
}

function cartasRestantes() {
  return shoe.length;
}

function barajasRestantes() {
  return shoe.length / 52;
}

// ============================================================
//  PUNTOS
//  El As vale 11, pero baja a 1 si te ayuda a no pasarte de 21.
// ============================================================
function calcularPuntos(cartas) {
  let total = 0;
  let ases = 0;
  for (const carta of cartas) {
    if (carta.valor === "A") {
      ases++;
      total += 11;
    } else if (["J", "Q", "K"].includes(carta.valor)) {
      total += 10;
    } else {
      total += Number(carta.valor);
    }
  }
  while (total > 21 && ases > 0) {
    total -= 10;
    ases--;
  }
  return total;
}

// Devuelve { total, suave } — "suave" = tiene un As contando como 11
function infoMano(cartas) {
  let total = 0;
  let ases = 0;
  for (const carta of cartas) {
    if (carta.valor === "A") {
      ases++;
      total += 11;
    } else if (["J", "Q", "K"].includes(carta.valor)) {
      total += 10;
    } else {
      total += Number(carta.valor);
    }
  }
  while (total > 21 && ases > 0) {
    total -= 10;
    ases--;
  }
  return { total, suave: ases > 0 };
}

// Valor a efectos de división: las figuras y el 10 cuentan igual ("10")
function valorSplit(carta) {
  if (carta.valor === "A") return "A";
  if (["10", "J", "Q", "K"].includes(carta.valor)) return "10";
  return carta.valor;
}
function mismoValorSplit(a, b) {
  return valorSplit(a) === valorSplit(b);
}

// ============================================================
//  CONTEO DE CARTAS Hi-Lo
//  2-6 suman +1, 7-9 valen 0, 10/figuras/As restan -1.
//  Se cuenta cada carta cuando se hace VISIBLE.
// ============================================================
function valorConteo(carta) {
  if (["2", "3", "4", "5", "6"].includes(carta.valor)) return 1;
  if (["7", "8", "9"].includes(carta.valor)) return 0;
  return -1; // 10, J, Q, K, A
}

function contar(carta) {
  conteoCorrido += valorConteo(carta);
}

// Conteo "verdadero" = corrido / barajas restantes (ajusta por tamaño del shoe)
function conteoVerdadero() {
  const baraj = barajasRestantes();
  if (baraj < 0.1) return conteoCorrido;
  return conteoCorrido / baraj;
}

// ============================================================
//  PROBABILIDAD DE PASARSE si pides otra carta
//  Mira las cartas que quedan en el shoe y cuenta cuántas te
//  harían superar 21.
// ============================================================
function probabilidadPasarse(cartas) {
  const total = calcularPuntos(cartas);
  if (total >= 21) return total > 21 ? 1 : 0;
  if (shoe.length === 0) return 0;
  let pasan = 0;
  for (const c of shoe) {
    if (calcularPuntos([...cartas, c]) > 21) pasan++;
  }
  return pasan / shoe.length;
}

File: lib/features/game/domain/cartas.dart
// ============================================================
//  Cartas, mazo múltiple (Shoe), puntos, conteo y probabilidades
//
//  Lógica pura traducida de la versión web (js/cartas.js). No
//  toca UI ni red. Reglas: docs/reglas-negocio/reglas-del-juego.md
// ============================================================

import 'dart:math';

import 'modelos.dart';

const List<String> _figuras = ['J', 'Q', 'K'];

/// Calcula los puntos de una mano. El As vale 11, pero baja a 1 las veces
/// necesarias para no pasarse de 21.
int calcularPuntos(List<Carta> cartas) {
  var total = 0;
  var ases = 0;
  for (final carta in cartas) {
    if (carta.valor == 'A') {
      ases++;
      total += 11;
    } else if (_figuras.contains(carta.valor)) {
      total += 10;
    } else {
      total += int.parse(carta.valor);
    }
  }
  while (total > 21 && ases > 0) {
    total -= 10;
    ases--;
  }
  return total;
}

/// Devuelve el total de la mano y si es "suave" (tiene un As contando como 11).
InfoMano infoMano(List<Carta> cartas) {
  var total = 0;
  var ases = 0;
  for (final carta in cartas) {
    if (carta.valor == 'A') {
      ases++;
      total += 11;
    } else if (_figuras.contains(carta.valor)) {
      total += 10;
    } else {
      total += int.parse(carta.valor);
    }
  }
  while (total > 21 && ases > 0) {
    total -= 10;
    ases--;
  }
  return InfoMano(total, ases > 0);
}

/// Valor a efectos de división: las figuras y el 10 cuentan igual ('10').
String valorSplit(Carta carta) {
  if (carta.valor == 'A') return 'A';
  if (carta.valor == '10' || _figuras.contains(carta.valor)) return '10';
  return carta.valor;
}

/// ¿Dos cartas tienen el mismo valor de división (se pueden partir)?
bool mismoValorSplit(Carta a, Carta b) => valorSplit(a) == valorSplit(b);

/// Valor Hi-Lo de una carta: 2–6 = +1, 7–9 = 0, 10/figuras/As = −1.
int valorConteo(Carta carta) {
  const bajas = ['2', '3', '4', '5', '6'];
  const neutras = ['7', '8', '9'];
  if (bajas.contains(carta.valor)) return 1;
  if (neutras.contains(carta.valor)) return 0;
  return -1; // 10, J, Q, K, A
}

/// Probabilidad de pasarse de 21 si se pide una carta más, dadas las cartas
/// que aún quedan disponibles ([restantes]).
double probabilidadPasarse(List<Carta> mano, List<Carta> restantes) {
  final total = calcularPuntos(mano);
  if (total >= 21) return total > 21 ? 1.0 : 0.0;
  if (restantes.isEmpty) return 0.0;
  var pasan = 0;
  for (final c in restantes) {
    if (calcularPuntos([...mano, c]) > 21) pasan++;
  }
  return pasan / restantes.length;
}

/// El "zapato": varias barajas mezcladas de las que se reparten cartas.
///
/// Mantiene el conteo Hi-Lo corrido. El barajado usa [Random], que se puede
/// inyectar (semilla fija) para tests deterministas.
class Shoe {
  final List<Carta> _cartas = [];
  final Random _random;

  int _totalInicial = 0;
  int _conteoCorrido = 0;

  /// Crea y baraja un shoe con [numBarajas] barajas.
  ///
  /// [random] permite fijar la semilla en tests; por defecto usa una aleatoria.
  Shoe(int numBarajas, {Random? random}) : _random = random ?? Random() {
    rebarajar(numBarajas);
  }

  /// Cartas que quedan por repartir.
  int get cartasRestantes => _cartas.length;

  /// Barajas que quedan (cartas restantes / 52).
  double get barajasRestantes => _cartas.length / 52;

  /// Conteo Hi-Lo corrido (running count).
  int get conteoCorrido => _conteoCorrido;

  /// Conteo "verdadero" = corrido / barajas restantes.
  double get conteoVerdadero {
    final baraj = barajasRestantes;
    if (baraj < 0.1) return _conteoCorrido.toDouble();
    return _conteoCorrido / baraj;
  }

  /// Vista de solo lectura de las cartas restantes (para cálculos como
  /// [probabilidadPasarse]).
  List<Carta> get restantes => List.unmodifiable(_cartas);

  /// ¿Conviene rebarajar? (penetración del 25%, como en los casinos).
  bool get necesitaBarajar => _cartas.length < _totalInicial * 0.25;

  /// Llena el shoe con [numBarajas] barajas y lo mezcla (Fisher-Yates).
  void rebarajar(int numBarajas) {
    _cartas.clear();
    for (var b = 0; b < numBarajas; b++) {
      for (final palo in Palo.values) {
        for (final valor in valoresBaraja) {
          _cartas.add(Carta(palo, valor));
        }
      }
    }
    _barajar();
    _totalInicial = _cartas.length;
    _conteoCorrido = 0;
  }

  /// Saca la carta de arriba del shoe. Si se acaba, lo regenera con el mismo
  /// número de barajas con el que se creó.
  Carta sacarCarta() {
    if (_cartas.isEmpty) {
      rebarajar(_totalInicial ~/ 52);
    }
    return _cartas.removeLast();
  }

  /// Suma una carta al conteo Hi-Lo (se llama al hacerla visible).
  void contar(Carta carta) {
    _conteoCorrido += valorConteo(carta);
  }

  void _barajar() {
    for (var i = _cartas.length - 1; i > 0; i--) {
      final j = _random.nextInt(i + 1);
      final tmp = _cartas[i];
      _cartas[i] = _cartas[j];
      _cartas[j] = tmp;
    }
  }
}
